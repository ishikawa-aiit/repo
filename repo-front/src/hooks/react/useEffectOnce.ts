import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
