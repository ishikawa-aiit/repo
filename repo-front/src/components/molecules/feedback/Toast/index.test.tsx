import { render, screen, act, within } from '@testing-library/react'

import { toast, Toast } from '.'

const body = 'サンプルボディ'
const status = {
  error: 'error' as const,
}
const title = 'サンプルタイトル'
const description = 'サンプルディスクリプション'

describe('props:children', () => {
  it('正常にレンダリングされている', () => {
    const { container } = render(<Toast>{body}</Toast>)

    const { getByText } = within(screen.getByTestId('Toast'))

    expect(getByText(body)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})

describe('props:title', () => {
  it('タイトルが正常にレンダリングされている', async () => {
    const { container } = render(<Toast>{body}</Toast>)
    const title = 'テストタイトル'

    await act(async () => {
      toast({
        status: status.error,
        title,
        description,
      })
    })

    const allByTitle = await screen.findByRole('alert', { name: title })

    expect(allByTitle).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})

describe('props:description', () => {
  it('ディスクリプションが正常にレンダリングされている', async () => {
    const { container } = render(<Toast>{body}</Toast>)
    const description = 'テストディスクリプション'

    await act(async () => {
      toast({
        status: status.error,
        title,
        description,
      })
    })

    const allByDescription = await screen.findByText(description)

    expect(allByDescription).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
