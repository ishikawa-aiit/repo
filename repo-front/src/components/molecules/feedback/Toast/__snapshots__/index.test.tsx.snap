// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props:children 正常にレンダリングされている 1`] = `
<div>
  <div
    class="emotion-0"
    data-cy="Toast"
    data-testid="Toast"
  >
    サンプルボディ
  </div>
</div>
`;

exports[`props:description ディスクリプションが正常にレンダリングされている 1`] = `
<div>
  <div
    class="emotion-0"
    data-cy="Toast"
    data-testid="Toast"
  >
    サンプルボディ
  </div>
</div>
`;

exports[`props:title タイトルが正常にレンダリングされている 1`] = `
<div>
  <div
    class="emotion-0"
    data-cy="Toast"
    data-testid="Toast"
  >
    サンプルボディ
  </div>
</div>
`;
